flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/AsyncExecutionController.java

```diff
@@ -0,0 +1,248 @@
+     * @return the state future.
+     */
+    public <IN, OUT> InternalStateFuture<OUT> handleRequest(
+            @Nullable State state, StateRequestType type, @Nullable IN payload) {
+        // Step 1: build state future & assign context.
+        InternalStateFuture<OUT> stateFuture = stateFutureFactory.build(currentContext);
+
+        StateRequest<K, IN, OUT> request =
+                new StateRequest<>(state, type, payload, stateFuture, currentContext);
+
+        // Step 2: try to seize the capacity, if the current in-flight records exceeds the limit,
+        // block the current state request from entering until some buffered requests are processed.
+        seizeCapacity();
+
+        // Step 3: try to occupy the key and place it into right buffer.
+        if (tryOccupyKey(currentContext)) {
+            insertActiveBuffer(request);
+        } else {
+            insertBlockingBuffer(request);
+        }
+        // Step 4: trigger the (active) buffer if needed.
+        triggerIfNeeded(false);
+        return stateFuture;
+    }
+
+    <IN, OUT> void insertActiveBuffer(StateRequest<K, IN, OUT> request) {
+        stateRequestsBuffer.enqueueToActive(request);
+    }
+
+    <IN, OUT> void insertBlockingBuffer(StateRequest<K, IN, OUT> request) {
+        stateRequestsBuffer.enqueueToBlocking(request);
+    }
+
+    /**
+     * Trigger a batch of requests.
+     *
+     * @param force whether to trigger requests in force.
+     */
+    void triggerIfNeeded(boolean force) {
+        if (!force && stateRequestsBuffer.activeQueueSize() < batchSize) {
+            return;
+        }
+        List<StateRequest<?, ?, ?>> toRun = stateRequestsBuffer.nextToRun();
+        stateRequestsBuffer.clearActiveQueue();
+        stateExecutor.executeBatchRequests(toRun);
+    }
+
+    private void seizeCapacity() {
+        // 1. Check if the record is already in buffer. If yes, this indicates that it is a state
+        // request resulting from a callback statement, otherwise, it signifies the initial state
+        // request
+        // for a newly entered record.
+        if (currentContext.isKeyOccupied()) {
+            return;
+        }
+        RecordContext<R, K> storedContext = currentContext;
+        // 2. If the state request is for a newly entered record, the in-flight record number should
+        // be less than the max in-flight record number.
+        // Note: the currentContext may be updated by {@code StateFutureFactory#build}.
+        try {
+            while (inFlightRecordNum.get() > maxInFlightRecordNum) {
+                if (!mailboxExecutor.tryYield()) {
+                    triggerIfNeeded(true);
+                    Thread.sleep(50);
+                }
+            }
+            // 3. Ensure the currentContext is restored.
+            setCurrentContext(storedContext);
+            inFlightRecordNum.incrementAndGet();
+        } catch (InterruptedException e) {
```

Comment:
IIUC, the `InterruptedException` should be ignored, otherwise it will produce a fatal error during TM normal exit (required by JM) ?

Reply:
👍good catch, `Thread.sleep(50);` is deleted, so the `InterruptedException` wouldn't throw.